snuba/state/rate_limit.py

```diff
@@ -0,0 +1,219 @@
+from collections import namedtuple
+from contextlib import contextmanager, AbstractContextManager, ExitStack
+from dataclasses import dataclass
+import logging
+import time
+from types import TracebackType
+from typing import (
+    Iterator,
+    Mapping,
+    MutableMapping,
+    Optional,
+    Sequence,
+    Type
+)
+import uuid
+
+from snuba import state
+
+logger = logging.getLogger('snuba.state.rate_limit')
+
+PROJECT_RATE_LIMIT_NAME = 'project'
+GLOBAL_RATE_LIMIT_NAME = 'global'
+
+
+@dataclass(frozen=True)
+class RateLimitParameters:
+    """
+    The configuration object which defines all the needed properties to create
+    a rate limit.
+    """
+
+    rate_limit_name: str
+    bucket: str
+    per_second_limit: Optional[float]
+    concurrent_limit: Optional[int]
+
+
+class RateLimitExceeded(Exception):
+    """
+    Exception thrown when the rate limit is exceeded
+    """
+
+
+@dataclass(frozen=True)
+class RateLimitStats:
+    """
+    The stats returned after the rate limit is run to tell the caller about the current
+    rate and number of concurrent requests.
+    """
+    rate: float
+    concurrent: int
+
+
+class RateLimitStatsContainer:
+    """
+    A container to collect stats for all the rate limits that have been run.
+    """
+    def __init__(self) -> None:
+        self.stats: MutableMapping[str, RateLimitStats] = {}
```

Comment:
if you are supposed to read these stats through the getter method. Please rename this to self.__stats